force-app/main/default/classes/DMLOperations.cls

```diff
@@ -128,92 +107,130 @@ public with sharing class DMLOperations {
-	 * If an Account with the given name already exists, update the description to "Updated Account".
-	 * If no such Account exists, it creates a new one and sets the description to "New Account".
+	 * If an Account with the given name already exists, it updates that Account.
+	 * If no such Account exists, it creates a new one.
 	 * @param accountName The name of the Account to be upserted.
 	 * @return The Account that was upserted.
 	 */
 	public static Account upsertAccount(String accountName) {
-		// Query for the Account
-		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
-
-		// if the Account exists, update the description
-
-		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
+        Account acc;
+		List<Account> accList = [SELECT Id, Name FROM Account WHERE Name =: accountName];
+        if(!accList.isEmpty()) {
+            acc = accList[0];
+            acc.Description = 'Updated Account';
+        } else {
+            acc = new Account(Name = accountName);
+        }
+        upsert acc;
+        return acc;
 	}
 
     /**
      * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
-     * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
-     * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
+	 * Upserts accounts and links contacts to them based on the contact's last name and account name.
+	 * Create or use existing account with contact last name. (Hint: use previous upsertAccount method)
+	 * Contacts are then linked to these accounts based on a full last name match.
+     * Example Contact Last Name = Doe Link to Account Name = Doe 
+     * Example Contact Last Name = Jane Link to Account Name = Jane
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+        String accountName;
+        List<Contact> contactsToUpdate = new List<Contact>();
+        for(Contact singleContact : contactsToUpsert) {
+            accountName = singleContact.LastName;
+            Account acc = upsertAccount(accountName);
+            singleContact.AccountId = acc.Id;
+            contactsToUpdate.add(singleContact);
+        }
+        upsert contactsToUpdate;
+    }
```

Comment:
Technically, you didn't need to create a new list (`contactsToUpdate`) since you already had `contactsToUpsert`. As a result, you could have simply used the DML statement on `contactsToUpsert`:
```suggestion
        String accountName;
        for(Contact singleContact : contactsToUpsert) {
            accountName = singleContact.LastName;
            Account acc = upsertAccount(accountName);
            singleContact.AccountId = acc.Id;
        }
        upsert contactsToUpsert;
    }
```
And not to confuse the issue, with this code we're actually updating only existing Accounts so the DML statement could be update as opposed to upsert. (This is because any new Accounts created were already inserted in the `upsertAccount()` method.)